pkg/cgroups/manager/manager.go

```diff
@@ -0,0 +1,392 @@
+		return
+	}
+	m.podEvents <- podEvent{
+		pod:       pod,
+		eventType: podAddEvent,
+	}
+}
+
+func (m *CgroupManager) OnUpdatePod(oldPod, newPod *v1.Pod) {
+	if newPod.Spec.NodeName != nodetypes.GetName() {
+		return
+	}
+	m.podEvents <- podEvent{
+		pod:       newPod,
+		oldPod:    oldPod,
+		eventType: podUpdateEvent,
+	}
+}
+
+func (m *CgroupManager) OnDeletePod(pod *v1.Pod) {
+	if pod.Spec.NodeName != nodetypes.GetName() {
+		return
+	}
+	m.podEvents <- podEvent{
+		pod:       pod,
+		eventType: podDeleteEvent,
+	}
+}
+
+// GetPodMetadataForContainer returns pod metadata for the given container
+// cgroup id in case of success, or nil otherwise.
+func (m *CgroupManager) GetPodMetadataForContainer(cgroupId uint64) *PodMetadata {
+	if !m.enabled {
+		return nil
+	}
+	podMetaOut := make(chan *PodMetadata)
+
+	m.podEvents <- podEvent{
+		cgroupId:       cgroupId,
+		eventType:      podGetMetadataEvent,
+		podMetadataOut: podMetaOut,
+	}
+	select {
+	// We either receive pod metadata, or zero value when the channel is closed.
+	case pm := <-podMetaOut:
+		return pm
+	}
+}
+
+// Close should only be called once from daemon close.
+func (m *CgroupManager) Close() {
+	close(m.shutdown)
+}
+
+type podUID = string
+
+type podMetadata struct {
+	name       string
+	namespace  string
+	ips        []string
+	containers map[string]struct{}
+}
+
+type containerMetadata struct {
+	cgroupId    uint64
+	cgroupPath  string
+	parentPodId string
+}
+
+type podEvent struct {
+	pod            *v1.Pod
+	oldPod         *v1.Pod
+	cgroupId       uint64
+	eventType      int
+	podMetadataOut chan *PodMetadata
+}
+
+type fs interface {
+	Stat(name string) (os.FileInfo, error)
+}
+
+type cgroup interface {
+	GetCgroupID(cgroupPath string) (uint64, error)
+}
+
+type fsImpl struct{}
+
+func (f fsImpl) Stat(name string) (os.FileInfo, error) {
+	return os.Stat(name)
+}
+
+type cgroupImpl struct{}
+
+func (c cgroupImpl) GetCgroupID(cgroupPath string) (uint64, error) {
+	return cgroups.GetCgroupID(cgroupPath)
+}
+
+func initManager(fs fs, cg cgroup, channelSize int) *CgroupManager {
+	m := &CgroupManager{
+		podMetadataById:           make(map[string]*podMetadata),
+		containerMetadataByCgrpId: make(map[uint64]*containerMetadata),
+		podEvents:                 make(chan podEvent, channelSize),
+		shutdown:                  make(chan struct{}),
+	}
+	m.fschecker = fs
+	m.cgroupsChecker = cg
+	m.checkCgroupPath = new(sync.Once)
+
+	m.enable()
+	go m.processPodEvents()
+
+	return m
+}
+
+func (m *CgroupManager) enable() {
+	if !option.Config.EnableSocketLBTracing {
+		m.enabled = false
+		return
+	}
+	m.enabled = true
+	m.checkCgroupPath.Do(func() {
+		for _, path := range cgroupBasePaths {
+			if _, err := m.fschecker.Stat(path); err != nil {
+				continue
+			}
+			m.templateCgroupBasePath = path
+			break
+		}
+		if m.templateCgroupBasePath == "" {
+			log.Warnf("No valid cgroup base path found: socket " +
+				"load-balancing tracing feature will not work")
```

Comment:
Couple of notes here:
- No need to use the formatting variation of the logging function if there is no formatting directives
- Should we suggest to the user to collect some specific information from their system and file an issue on github? If they hit this, we will want to know what the real cgroup path is.
```suggestion
			log.Warn("No valid cgroup base path found: socket " +
				"load-balancing tracing feature will not work")
```

Reply:
>Should we suggest to the user to collect some specific information from their system and file an issue on github? If they hit this, we will want to know what the real cgroup path is.

Yeah, I was going to put this in documentation so that we can mention what the current defaults are. But having it logged as part of the warning is fine too, we can add more details to the documentation.